modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java

```diff
@@ -0,0 +1,424 @@
+    public boolean wasApplied() {
+        if (hasRowSet() || cur.queryType() == SqlQueryType.DML) {
+            return false;
+        }
+
+        assert batchPage.items().size() == 1
+                && batchPage.items().get(0).size() == 1
+                && batchPage.items().get(0).get(0) instanceof Boolean
+                && !batchPage.hasMore() : "Invalid DDL result: " + batchPage;
+
+        return (boolean) batchPage.items().get(0).get(0);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Iterable<SqlRow> currentPage() {
+        if (!hasRowSet()) {
+            throw new NoRowSetExpectedException("Query hasn't result set: [type=" + cur.queryType() + ']');
+        }
+
+        return page;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletionStage<? extends AsyncResultSet> fetchNextPage() {
+        if (next == null) {
+            synchronized (mux) {
+                if (next == null) {
+                    if (!hasMorePages()) {
+                        next = HAS_NO_MORE_PAGE_FUTURE;
+                    } else {
+                        next = cur.requestNextAsync(pageSize)
+                                .thenApply(batchRes -> new AsyncResultSetImpl(cur, batchRes, pageSize, closeRun));
+                    }
+                }
+            }
+        }
+
+        return next;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean hasMorePages() {
+        return batchPage.hasMore();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletionStage<Void> closeAsync() {
+        return cur.closeAsync().thenAccept((v) -> closeRun.run());
+    }
+
+    private class Page implements Iterable<SqlRow> {
+        /** {@inheritDoc} */
+        @NotNull
+        @Override
+        public Iterator<SqlRow> iterator() {
+            return new IteratorImpl(batchPage.items().iterator());
+        }
+    }
+
+    private class IteratorImpl implements Iterator<SqlRow> {
+        private final Iterator<List<Object>> it;
+
+        IteratorImpl(Iterator<List<Object>> it) {
+            this.it = it;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return it.hasNext();
+        }
+
+        @Override
+        public SqlRow next() {
+            return new SqlRowImpl(it.next());
+        }
+    }
+
+    private class SqlRowImpl implements SqlRow {
+        private final List<Object> row;
+
+        SqlRowImpl(List<Object> row) {
+            this.row = row;
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public int columnCount() {
+            return cur.metadata().fields().size();
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public String columnName(int columnIndex) {
+            return cur.metadata().fields().get(columnIndex).name();
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public int columnIndex(@NotNull String columnName) {
+            return cur.metadata().fields().stream()
```

Comment:
does it make sense to prepare metadata once for the whole result and then just pass it to the row?

BTW, what if the column with the given name is not exists in the row?